src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs

```diff
@@ -326,10 +316,30 @@ void WriteToChannel(ResourceEvent resourceEvent) =>
             OnResourceUpdated += WriteToChannel;
         }
 
+        long id = 0;
+
+        // Return the last snapshot for each resource.
+        foreach (var state in _resourceNotificationStates)
+        {
+            var (resource, resourceId) = state.Key;
+
+            if (state.Value.LastSnapshot is { } ss)
+            {
+                id = Math.Max(id, ss.Id);
+
+                yield return new ResourceEvent(resource, resourceId, state.Value.LastSnapshot);
+            }
+        }
+
         try
         {
             await foreach (var item in channel.Reader.ReadAllAsync(cancellationToken).ConfigureAwait(false))
             {
+                if (item.Snapshot.Id <= id)
```

Comment:
When could this happen? Multiple publishes from different threads?

If this is unexpected, should a log be written?

Reply:
It's something we can't avoid without more synchronization (which I am trying to avoid), so it isn't worth logging (well maybe in trace we can?). We're missing events in some cases, so this is trying to make sure we don't get old events.

Reply:
How does that happen? PublishUpdateAsync happens in a lock, which writes to the happen in order, and the channel reads in order. When do things go out of order?

Reply:
The lock is per resource, not globally. We return the last snapshot before we subscribe for updates, in between those 2 operations where we can drop new publishes.